### DISCUSSION TITLE
What is the URL format for api.jspm.io/generate?

### DISCUSSION THREAD

**Discussion by trusktr** [jspm/project#307] [Q&A]
I am desiring to make concise for easy pasting to get libraries "installed" for build-less users. The import maps generated by the generator work, but they can be a lot of noisy code.

I'd like to have this instead:

```html
<script src="path/to/import-map.js" for="library@1.2.3,other-library@1.2.3"></script>
```

and it will sync XHR the importmap from JSPM, then `document.write` it, to keep the code short.

I can implement this, but I don't know what the format is for https://api.jspm.io/generate. Can you briefly summarize the URL format(s)?

**Comment by trusktr** [comment]
I found details on the format here:

- https://blog.appsignal.com/2022/03/02/import-maps-under-the-hood-in-rails-7.html
- https://blog.tunaxor.me/blog/2021-12-16-Building-a-webpack-alternative-in-fsharp.html

Example request:

https://api.jspm.io/generate?install=lume

Play with it live here:

https://hopp.sh/r/1jvo7XXmC8gM

It accepts the following parameters (not sure this is correct):

- **install** - `["jquery"]` - the param is an array because you can install multiple packages simultaneously.
- **flattenScope** - `true` - with this param set to true, the returned import map format will be more straightforward without scopes, if possible.
- **env** - `["browser", "module", "production"]` - this is the list of environment condition strings.
- **provider** - `"jspm"` - besides JSPM, Skypack, JSdelivr, and Unpkg providers are available to use as well.

```js
const response = await fetch('https://api.jspm.io/generate?install=lume')
const info = await response.json()

console.log(info.map) // the importmap
```

The response takes about 2 seconds for me every time though. It is not cached. Is there a way to cache it? Maybe by providing a cache key parameter or something?

**Reply by guybedford** [reply]
Generator API documentation is located at https://jspm.org/cdn/api#generator.

All versions need to be looked up every time a generate call is run. Caching layers would be the same as any caching in the main generator project itself (https://github.com/jspm/generator).

The source for this documentation is at https://github.com/jspm/jspm.org/blob/main/pages/cdn/api.md, if you have any suggestions for improvements or clarifications.

**Reply by trusktr** [reply]
Yeah I saw the docs, which state "use GET or POST", but they didn't state anywhere what to actually send to the endpoint. I can't really improve the docs without knowing the details of the server.

What is missing from above that I can't possibly infer from the private undocumented generator endpoint?

**Reply by guybedford** [reply]
Thanks, I've added a note to this feedback in https://github.com/jspm/jspm.org/issues/44.
